File: libs/ontology-kernel/src/models/action_type.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq, Eq)]
#[sqlx(type_name = "text", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum ActionOperationKind {
    UpdateObject,
    CreateLink,
    DeleteObject,
    InvokeFunction,
    InvokeWebhook,
    /// TASK I — Create an object that implements the configured interface.
    /// Resolves to a concrete `object_type_id` at runtime via the auto-
    /// generated `__object_type` parameter.
    CreateInterface,
    /// TASK I — Modify an existing object referenced via `__interface_ref`,
    /// resolving its concrete `object_type_id` at runtime.
    ModifyInterface,
    /// TASK I — Delete an existing object referenced via `__interface_ref`.
    DeleteInterface,
    /// TASK I — Create a link between two interface-typed endpoints.
    CreateInterfaceLink,
    /// TASK I — Delete an existing link between two interface-typed endpoints.
    DeleteInterfaceLink,
}

impl std::fmt::Display for ActionOperationKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let value = match self {
            Self::UpdateObject => "update_object",
            Self::CreateLink => "create_link",
            Self::DeleteObject => "delete_object",
            Self::InvokeFunction => "invoke_function",
            Self::InvokeWebhook => "invoke_webhook",
            Self::CreateInterface => "create_interface",
            Self::ModifyInterface => "modify_interface",
            Self::DeleteInterface => "delete_interface",
            Self::CreateInterfaceLink => "create_interface_link",
            Self::DeleteInterfaceLink => "delete_interface_link",
        };

        write!(f, "{value}")
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionInputField {
    pub name: String,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub property_type: String,
    #[serde(default)]
    pub required: bool,
    pub default_value: Option<Value>,
    /// TASK J — Struct parameters carry nested fields. Only meaningful when
    /// `property_type == "struct"`. Each nested field follows the same
    /// `ActionInputField` shape (recursive). Other property types must leave
    /// this empty/None.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub struct_fields: Option<Vec<ActionInputField>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionPropertyMapping {
    pub property_name: String,
    pub input_name: Option<String>,
    pub value: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateObjectActionConfig {
    pub property_mappings: Vec<ActionPropertyMapping>,
    #[serde(default)]
    pub static_patch: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct ActionFormSchema {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub sections: Vec<ActionFormSection>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub parameter_overrides: Vec<ActionFormParameterOverride>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ActionFormSection {
    pub id: String,
    pub title: Option<String>,
    pub description: Option<String>,
    #[serde(default = "default_action_form_section_columns")]
    pub columns: u8,
    #[serde(default)]
    pub collapsible: bool,
    #[serde(default = "default_action_form_section_visible")]
    pub visible: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub parameter_names: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub overrides: Vec<ActionFormSectionOverride>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ActionFormSectionOverride {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub conditions: Vec<ActionFormCondition>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hidden: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub columns: Option<u8>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ActionFormParameterOverride {
    pub parameter_name: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub conditions: Vec<ActionFormCondition>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub hidden: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub required: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub default_value: Option<Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ActionFormCondition {
    pub left: String,
    #[serde(default = "default_action_form_condition_operator")]
    pub operator: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub right: Option<Value>,
}

fn default_action_form_section_columns() -> u8 {
    1
}

fn default_action_form_section_visible() -> bool {
    true
}

fn default_action_form_condition_operator() -> String {
    "is".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ActionAuthorizationPolicy {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub required_permission_keys: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub any_role: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub all_roles: Vec<String>,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub attribute_equals: HashMap<String, Value>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub allowed_markings: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub minimum_clearance: Option<String>,
    #[serde(default)]
    pub deny_guest_sessions: bool,
}

#[derive(Debug, Clone, FromRow)]
pub struct ActionTypeRow {
    pub id: Uuid,
    pub name: String,
    pub display_name: String,
    pub description: String,
    pub object_type_id: Uuid,
    pub operation_kind: String,
    pub input_schema: Value,
    pub form_schema: Value,
    pub config: Value,
    pub confirmation_required: bool,
    pub permission_key: Option<String>,
    pub authorization_policy: Value,
    pub owner_id: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionType {
    pub id: Uuid,
    pub name: String,
    pub display_name: String,
    pub description: String,
    pub object_type_id: Uuid,
    pub operation_kind: String,
    pub input_schema: Vec<ActionInputField>,
    pub form_schema: ActionFormSchema,
    pub config: Value,
    pub confirmation_required: bool,
    pub permission_key: Option<String>,
    pub authorization_policy: ActionAuthorizationPolicy,
    pub owner_id: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl TryFrom<ActionTypeRow> for ActionType {
    type Error = serde_json::Error;

    fn try_from(row: ActionTypeRow) -> Result<Self, Self::Error> {
        Ok(Self {
            id: row.id,
            name: row.name,
            display_name: row.display_name,
            description: row.description,
            object_type_id: row.object_type_id,
            operation_kind: row.operation_kind,
            input_schema: serde_json::from_value(row.input_schema).unwrap_or_default(),
            form_schema: serde_json::from_value(row.form_schema).unwrap_or_default(),
            config: row.config,
            confirmation_required: row.confirmation_required,
            permission_key: row.permission_key,
            authorization_policy: serde_json::from_value(row.authorization_policy)
                .unwrap_or_default(),
            owner_id: row.owner_id,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }
}

#[derive(Debug, Deserialize)]
pub struct CreateActionTypeRequest {
    pub name: String,
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub object_type_id: Uuid,
    pub operation_kind: String,
    pub input_schema: Option<Vec<ActionInputField>>,
    pub form_schema: Option<ActionFormSchema>,
    pub config: Option<Value>,
    pub confirmation_required: Option<bool>,
    pub permission_key: Option<String>,
    pub authorization_policy: Option<ActionAuthorizationPolicy>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateActionTypeRequest {
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub operation_kind: Option<String>,
    pub input_schema: Option<Vec<ActionInputField>>,
    pub form_schema: Option<ActionFormSchema>,
    pub config: Option<Value>,
    pub confirmation_required: Option<bool>,
    pub permission_key: Option<String>,
    pub authorization_policy: Option<ActionAuthorizationPolicy>,
}

#[derive(Debug, Deserialize)]
pub struct ListActionTypesQuery {
    pub object_type_id: Option<Uuid>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
    pub search: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ListActionTypesResponse {
    pub data: Vec<ActionType>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
}

#[derive(Debug, Deserialize)]
pub struct ValidateActionRequest {
    pub target_object_id: Option<Uuid>,
    #[serde(default)]
    pub parameters: Value,
}

#[derive(Debug, Serialize)]
pub struct ValidateActionResponse {
    pub valid: bool,
    pub errors: Vec<String>,
    pub preview: Value,
}

#[derive(Debug, Deserialize)]
pub struct ExecuteActionRequest {
    pub target_object_id: Option<Uuid>,
    #[serde(default)]
    pub parameters: Value,
    pub justification: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ExecuteBatchActionRequest {
    pub target_object_ids: Vec<Uuid>,
    #[serde(default)]
    pub parameters: Value,
    pub justification: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ExecuteActionResponse {
    pub action: ActionType,
    pub target_object_id: Option<Uuid>,
    pub deleted: bool,
    pub preview: Value,
    pub object: Option<Value>,
    pub link: Option<Value>,
    pub result: Option<Value>,
}

#[derive(Debug, Serialize)]
pub struct ExecuteBatchActionResponse {
    pub action: ActionType,
    pub total: usize,
    pub succeeded: usize,
    pub failed: usize,
    pub results: Vec<Value>,
}

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct ActionWhatIfBranch {
    pub id: Uuid,
    pub action_id: Uuid,
    pub target_object_id: Option<Uuid>,
    pub name: String,
    pub description: String,
    pub parameters: Value,
    pub preview: Value,
    pub before_object: Option<Value>,
    pub after_object: Option<Value>,
    pub deleted: bool,
    pub owner_id: Uuid,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct CreateActionWhatIfBranchRequest {
    pub target_object_id: Option<Uuid>,
    #[serde(default)]
    pub parameters: Value,
    pub name: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ListActionWhatIfBranchesQuery {
    pub target_object_id: Option<Uuid>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

#[derive(Debug, Serialize)]
pub struct ListActionWhatIfBranchesResponse {
    pub data: Vec<ActionWhatIfBranch>,
    pub total: i64,
    pub page: i64,
    pub per_page: i64,
}

File: libs/ontology-kernel/models/action_type.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// ActionOperationKind mirrors `enum ActionOperationKind` in
// `libs/ontology-kernel/src/models/action_type.rs`. Every variant
// serialises to its snake_case spelling for both serde and sqlx
// (TASK I additions are present verbatim).
type ActionOperationKind string

const (
	ActionOperationKindUpdateObject        ActionOperationKind = "update_object"
	ActionOperationKindCreateLink          ActionOperationKind = "create_link"
	ActionOperationKindDeleteObject        ActionOperationKind = "delete_object"
	ActionOperationKindInvokeFunction      ActionOperationKind = "invoke_function"
	ActionOperationKindInvokeWebhook       ActionOperationKind = "invoke_webhook"
	ActionOperationKindCreateInterface     ActionOperationKind = "create_interface"
	ActionOperationKindModifyInterface     ActionOperationKind = "modify_interface"
	ActionOperationKindDeleteInterface     ActionOperationKind = "delete_interface"
	ActionOperationKindCreateInterfaceLink ActionOperationKind = "create_interface_link"
	ActionOperationKindDeleteInterfaceLink ActionOperationKind = "delete_interface_link"
)

// String mirrors `impl Display for ActionOperationKind`.
func (k ActionOperationKind) String() string { return string(k) }

// ActionInputField mirrors `struct ActionInputField`. `required` is
// `#[serde(default)]` (false when missing), `struct_fields` is
// `skip_serializing_if = "Option::is_none"` so absent / nil
// pointers omit the key entirely.
type ActionInputField struct {
	Name         string             `json:"name"`
	DisplayName  *string            `json:"display_name"`
	Description  *string            `json:"description"`
	PropertyType string             `json:"property_type"`
	Required     bool               `json:"required"`
	DefaultValue json.RawMessage    `json:"default_value"`
	StructFields *[]ActionInputField `json:"struct_fields,omitempty"`
}

// ActionPropertyMapping mirrors `struct ActionPropertyMapping`.
type ActionPropertyMapping struct {
	PropertyName string          `json:"property_name"`
	InputName    *string         `json:"input_name"`
	Value        json.RawMessage `json:"value"`
}

// UpdateObjectActionConfig mirrors `struct UpdateObjectActionConfig`.
// `static_patch` is `#[serde(default)]` Option<Value>` — null when
// missing, encoded as `null` when nil.
type UpdateObjectActionConfig struct {
	PropertyMappings []ActionPropertyMapping `json:"property_mappings"`
	StaticPatch      json.RawMessage         `json:"static_patch"`
}

// ActionFormSchema mirrors `struct ActionFormSchema`. Both inner
// vecs carry `skip_serializing_if = "Vec::is_empty"`.
type ActionFormSchema struct {
	Sections           []ActionFormSection           `json:"sections,omitempty"`
	ParameterOverrides []ActionFormParameterOverride `json:"parameter_overrides,omitempty"`
}

// ActionFormSection mirrors `struct ActionFormSection`. Custom defaults:
//   * `columns` defaults to 1 (`default_action_form_section_columns`).
//   * `visible` defaults to true (`default_action_form_section_visible`).
//   * `parameter_names`, `overrides` default to empty Vec.
//   * Empty Vec fields skip on serialise.
type ActionFormSection struct {
	ID             string                       `json:"id"`
	Title          *string                      `json:"title,omitempty"`
	Description    *string                      `json:"description,omitempty"`
	Columns        uint8                        `json:"columns"`
	Collapsible    bool                         `json:"collapsible"`
	Visible        bool                         `json:"visible"`
	ParameterNames []string                     `json:"parameter_names,omitempty"`
	Overrides      []ActionFormSectionOverride  `json:"overrides,omitempty"`
}

// UnmarshalJSON applies `default_action_form_section_columns()` and
// `default_action_form_section_visible()` for missing keys.
func (s *ActionFormSection) UnmarshalJSON(b []byte) error {
	type alias ActionFormSection
	a := alias{Columns: 1, Visible: true}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(b, &raw); err != nil {
		return err
	}
	if err := json.Unmarshal(b, &a); err != nil {
		return err
	}
	if _, ok := raw["columns"]; !ok {
		a.Columns = 1
	}
	if _, ok := raw["visible"]; !ok {
		a.Visible = true
	}
	*s = ActionFormSection(a)
	return nil
}

// ActionFormSectionOverride mirrors `struct ActionFormSectionOverride`.
type ActionFormSectionOverride struct {
	Conditions  []ActionFormCondition `json:"conditions,omitempty"`
	Hidden      *bool                 `json:"hidden,omitempty"`
	Columns     *uint8                `json:"columns,omitempty"`
	Title       *string               `json:"title,omitempty"`
	Description *string               `json:"description,omitempty"`
}

// ActionFormParameterOverride mirrors `struct ActionFormParameterOverride`.
type ActionFormParameterOverride struct {
	ParameterName string                `json:"parameter_name"`
	Conditions    []ActionFormCondition `json:"conditions,omitempty"`
	Hidden        *bool                 `json:"hidden,omitempty"`
	Required      *bool                 `json:"required,omitempty"`
	DefaultValue  json.RawMessage       `json:"default_value,omitempty"`
	DisplayName   *string               `json:"display_name,omitempty"`
	Description   *string               `json:"description,omitempty"`
}

// ActionFormCondition mirrors `struct ActionFormCondition`. `operator`
// defaults to `"is"` (`default_action_form_condition_operator`).
type ActionFormCondition struct {
	Left     string          `json:"left"`
	Operator string          `json:"operator"`
	Right    json.RawMessage `json:"right,omitempty"`
}

// UnmarshalJSON applies `default_action_form_condition_operator()`.
func (c *ActionFormCondition) UnmarshalJSON(b []byte) error {
	type alias ActionFormCondition
	a := alias{Operator: "is"}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(b, &raw); err != nil {
		return err
	}
	if err := json.Unmarshal(b, &a); err != nil {
		return err
	}
	if _, ok := raw["operator"]; !ok {
		a.Operator = "is"
	}
	*c = ActionFormCondition(a)
	return nil
}

// ActionAuthorizationPolicy mirrors `struct ActionAuthorizationPolicy`.
// All Vec / HashMap fields skip when empty, matching serde.
type ActionAuthorizationPolicy struct {
	RequiredPermissionKeys []string                   `json:"required_permission_keys,omitempty"`
	AnyRole                []string                   `json:"any_role,omitempty"`
	AllRoles               []string                   `json:"all_roles,omitempty"`
	AttributeEquals        map[string]json.RawMessage `json:"attribute_equals,omitempty"`
	AllowedMarkings        []string                   `json:"allowed_markings,omitempty"`
	MinimumClearance       *string                    `json:"minimum_clearance,omitempty"`
	DenyGuestSessions      bool                       `json:"deny_guest_sessions"`
}

// ActionTypeRow mirrors `struct ActionTypeRow`.
type ActionTypeRow struct {
	ID                   uuid.UUID       `db:"id"`
	Name                 string          `db:"name"`
	DisplayName          string          `db:"display_name"`
	Description          string          `db:"description"`
	ObjectTypeID         uuid.UUID       `db:"object_type_id"`
	OperationKind        string          `db:"operation_kind"`
	InputSchema          json.RawMessage `db:"input_schema"`
	FormSchema           json.RawMessage `db:"form_schema"`
	Config               json.RawMessage `db:"config"`
	ConfirmationRequired bool            `db:"confirmation_required"`
	PermissionKey        *string         `db:"permission_key"`
	AuthorizationPolicy  json.RawMessage `db:"authorization_policy"`
	OwnerID              uuid.UUID       `db:"owner_id"`
	CreatedAt            time.Time       `db:"created_at"`
	UpdatedAt            time.Time       `db:"updated_at"`
}

// ActionType mirrors `struct ActionType`.
type ActionType struct {
	ID                   uuid.UUID                 `json:"id"`
	Name                 string                    `json:"name"`
	DisplayName          string                    `json:"display_name"`
	Description          string                    `json:"description"`
	ObjectTypeID         uuid.UUID                 `json:"object_type_id"`
	OperationKind        string                    `json:"operation_kind"`
	InputSchema          []ActionInputField        `json:"input_schema"`
	FormSchema           ActionFormSchema          `json:"form_schema"`
	Config               json.RawMessage           `json:"config"`
	ConfirmationRequired bool                      `json:"confirmation_required"`
	PermissionKey        *string                   `json:"permission_key"`
	AuthorizationPolicy  ActionAuthorizationPolicy `json:"authorization_policy"`
	OwnerID              uuid.UUID                 `json:"owner_id"`
	CreatedAt            time.Time                 `json:"created_at"`
	UpdatedAt            time.Time                 `json:"updated_at"`
}

// IntoAction mirrors `TryFrom<ActionTypeRow> for ActionType`. Per
// Rust, JSON parse failures fall back to defaults
// (`unwrap_or_default()`).
func (row ActionTypeRow) IntoAction() ActionType {
	var input []ActionInputField
	if len(row.InputSchema) > 0 {
		_ = json.Unmarshal(row.InputSchema, &input)
	}
	var form ActionFormSchema
	if len(row.FormSchema) > 0 {
		_ = json.Unmarshal(row.FormSchema, &form)
	}
	var policy ActionAuthorizationPolicy
	if len(row.AuthorizationPolicy) > 0 {
		_ = json.Unmarshal(row.AuthorizationPolicy, &policy)
	}
	return ActionType{
		ID:                   row.ID,
		Name:                 row.Name,
		DisplayName:          row.DisplayName,
		Description:          row.Description,
		ObjectTypeID:         row.ObjectTypeID,
		OperationKind:        row.OperationKind,
		InputSchema:          input,
		FormSchema:           form,
		Config:               row.Config,
		ConfirmationRequired: row.ConfirmationRequired,
		PermissionKey:        row.PermissionKey,
		AuthorizationPolicy:  policy,
		OwnerID:              row.OwnerID,
		CreatedAt:            row.CreatedAt,
		UpdatedAt:            row.UpdatedAt,
	}
}

// CreateActionTypeRequest mirrors `struct CreateActionTypeRequest`.
type CreateActionTypeRequest struct {
	Name                 string                     `json:"name"`
	DisplayName          *string                    `json:"display_name,omitempty"`
	Description          *string                    `json:"description,omitempty"`
	ObjectTypeID         uuid.UUID                  `json:"object_type_id"`
	OperationKind        string                     `json:"operation_kind"`
	InputSchema          *[]ActionInputField        `json:"input_schema,omitempty"`
	FormSchema           *ActionFormSchema          `json:"form_schema,omitempty"`
	Config               json.RawMessage            `json:"config,omitempty"`
	ConfirmationRequired *bool                      `json:"confirmation_required,omitempty"`
	PermissionKey        *string                    `json:"permission_key,omitempty"`
	AuthorizationPolicy  *ActionAuthorizationPolicy `json:"authorization_policy,omitempty"`
}

// UpdateActionTypeRequest mirrors `struct UpdateActionTypeRequest`.
type UpdateActionTypeRequest struct {
	DisplayName          *string                    `json:"display_name,omitempty"`
	Description          *string                    `json:"description,omitempty"`
	OperationKind        *string                    `json:"operation_kind,omitempty"`
	InputSchema          *[]ActionInputField        `json:"input_schema,omitempty"`
	FormSchema           *ActionFormSchema          `json:"form_schema,omitempty"`
	Config               json.RawMessage            `json:"config,omitempty"`
	ConfirmationRequired *bool                      `json:"confirmation_required,omitempty"`
	PermissionKey        *string                    `json:"permission_key,omitempty"`
	AuthorizationPolicy  *ActionAuthorizationPolicy `json:"authorization_policy,omitempty"`
}

// ListActionTypesQuery mirrors `struct ListActionTypesQuery`.
type ListActionTypesQuery struct {
	ObjectTypeID *uuid.UUID `json:"object_type_id,omitempty"`
	Page         *int64     `json:"page,omitempty"`
	PerPage      *int64     `json:"per_page,omitempty"`
	Search       *string    `json:"search,omitempty"`
}

// ListActionTypesResponse mirrors `struct ListActionTypesResponse`.
type ListActionTypesResponse struct {
	Data    []ActionType `json:"data"`
	Total   int64        `json:"total"`
	Page    int64        `json:"page"`
	PerPage int64        `json:"per_page"`
}

// ValidateActionRequest mirrors `struct ValidateActionRequest`.
// `parameters` is `#[serde(default)]` Value.
type ValidateActionRequest struct {
	TargetObjectID *uuid.UUID      `json:"target_object_id,omitempty"`
	Parameters     json.RawMessage `json:"parameters"`
}

// ValidateActionResponse mirrors `struct ValidateActionResponse`.
type ValidateActionResponse struct {
	Valid   bool            `json:"valid"`
	Errors  []string        `json:"errors"`
	Preview json.RawMessage `json:"preview"`
}

// ExecuteActionRequest mirrors `struct ExecuteActionRequest`.
type ExecuteActionRequest struct {
	TargetObjectID *uuid.UUID      `json:"target_object_id,omitempty"`
	Parameters     json.RawMessage `json:"parameters"`
	Justification  *string         `json:"justification,omitempty"`
}

// ExecuteBatchActionRequest mirrors `struct ExecuteBatchActionRequest`.
type ExecuteBatchActionRequest struct {
	TargetObjectIDs []uuid.UUID     `json:"target_object_ids"`
	Parameters      json.RawMessage `json:"parameters"`
	Justification   *string         `json:"justification,omitempty"`
}

// ExecuteActionResponse mirrors `struct ExecuteActionResponse`.
type ExecuteActionResponse struct {
	Action         ActionType      `json:"action"`
	TargetObjectID *uuid.UUID      `json:"target_object_id"`
	Deleted        bool            `json:"deleted"`
	Preview        json.RawMessage `json:"preview"`
	Object         json.RawMessage `json:"object"`
	Link           json.RawMessage `json:"link"`
	Result         json.RawMessage `json:"result"`
}

// ExecuteBatchActionResponse mirrors `struct ExecuteBatchActionResponse`.
type ExecuteBatchActionResponse struct {
	Action    ActionType        `json:"action"`
	Total     int               `json:"total"`
	Succeeded int               `json:"succeeded"`
	Failed    int               `json:"failed"`
	Results   []json.RawMessage `json:"results"`
}

// ActionWhatIfBranch mirrors `struct ActionWhatIfBranch`.
type ActionWhatIfBranch struct {
	ID             uuid.UUID       `json:"id"               db:"id"`
	ActionID       uuid.UUID       `json:"action_id"        db:"action_id"`
	TargetObjectID *uuid.UUID      `json:"target_object_id" db:"target_object_id"`
	Name           string          `json:"name"             db:"name"`
	Description    string          `json:"description"      db:"description"`
	Parameters     json.RawMessage `json:"parameters"       db:"parameters"`
	Preview        json.RawMessage `json:"preview"          db:"preview"`
	BeforeObject   json.RawMessage `json:"before_object"    db:"before_object"`
	AfterObject    json.RawMessage `json:"after_object"     db:"after_object"`
	Deleted        bool            `json:"deleted"          db:"deleted"`
	OwnerID        uuid.UUID       `json:"owner_id"         db:"owner_id"`
	CreatedAt      time.Time       `json:"created_at"       db:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"       db:"updated_at"`
}

// CreateActionWhatIfBranchRequest mirrors `struct CreateActionWhatIfBranchRequest`.
type CreateActionWhatIfBranchRequest struct {
	TargetObjectID *uuid.UUID      `json:"target_object_id,omitempty"`
	Parameters     json.RawMessage `json:"parameters"`
	Name           *string         `json:"name,omitempty"`
	Description    *string         `json:"description,omitempty"`
}

// ListActionWhatIfBranchesQuery mirrors `struct ListActionWhatIfBranchesQuery`.
type ListActionWhatIfBranchesQuery struct {
	TargetObjectID *uuid.UUID `json:"target_object_id,omitempty"`
	Page           *int64     `json:"page,omitempty"`
	PerPage        *int64     `json:"per_page,omitempty"`
}

// ListActionWhatIfBranchesResponse mirrors `struct ListActionWhatIfBranchesResponse`.
type ListActionWhatIfBranchesResponse struct {
	Data    []ActionWhatIfBranch `json:"data"`
	Total   int64                `json:"total"`
	Page    int64                `json:"page"`
	PerPage int64                `json:"per_page"`
}
